docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+# ADR 082: Data Companion API
+
+## Changelog
+
+- 2022-09-10: First draft (@thanethomson)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes to effectively offload some data storage responsibilities and
+functionality to **single external companion service** in order to:
+
+1. Improve core consensus stability.
+2. Eventually reduce the amount of functionality for which the core team is
+   responsible, thereby improving maintainability.
+3. Still cater for certain use cases that require this data and functionality.
+
+The way the system is currently built is such that a Tendermint node is mostly
```

Comment:
This was fed by a belief that all users would run their own node and use CLI. I fought for thinking web-scale and hosted servers, and using rpc/apis with client libs as first class citizens. This first materialised in the LCD, which was a minimal first step, and I think web-first only really came into being with CosmJS and tools like telescope.

Before making any changes in the direction the ADR suggests, we need to older maintainers, especially @ebuchman , to agree that ease of self-hosting is a secondary goal, and the primary goal is scalability, stability, and features when managed by professional DevOps.

Once there is full agreement on that, we can work on a very clear design for that goal. If that is not supported, and self-hosting is considered important, then that should be clear. Working out that fundamental design decision via various details in this ADR leads to a compromise that does nothing right.

Reply:
If there is clear agreement on a radical change of tendermint and sdk to web-scale, then I would be happy to join some discussion. I know @tac0turtle has done a lot of work and thinking in that direction.

Reply:
> Before making any changes in the direction the ADR suggests, we need to older maintainers, especially @ebuchman , to agree that ease of self-hosting is a secondary goal, and the primary goal is scalability, stability, and features when managed by professional DevOps.

Could you please elaborate on how you see self-hosting to be a contrasting goal to management by professional DevOps? Can professional a DevOps team not self-host, for instance? Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

> If there is clear agreement on a radical change of tendermint and sdk to web-scale, then I would be happy to join some discussion.

Given that "web-scale" is a somewhat overloaded term and can cause confusion in terms of expectations if we were to make it a goal, I think it would be preferable to focus on solving specific problems. In this case, if users do want to scale their RPC independently of their full node/validator, then this approach gives them the option.

Are there other aspects of the system that you would want to see scaled out independently of the core consensus functionality? We'd be more than happy to discuss synchronously sometime.

Reply:
> Could you please elaborate on how you see self-hosting to be a contrasting goal to management by professional DevOps? Can professional a DevOps team not self-host, for instance? Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

+1 — "professional DevOps" teams are going to "self-host" nodes by definition, right? 

Maybe the distinction you're trying to draw here, @ethanfrey, is between professional/organizational operators and hobbyist/individual operators? If so, I'm on board. Self-hosting by hobbyist or individual operators is hopefully a secondary, nice-to-have goal of the project; self-hosting by professional or organizational operators is hopefully a primary and mandatory goal of the project.

Reply:
Afaict, catering to self-hosting by professional/highly skilled/organizational operators of validators is definitely more important than facilitating that for hobbyists, unless someone has a really compelling argument otherwise.

Reply:
> Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

Yes, the primary focus was on hobbyists running a full node (not a validator) with minimal work. 

This idea seems to be about breaking it into multiple services to allow faster development speed, use of optional services not needed on validator nodes, more features and less attack vectors on (minimal) validator nodes.

I see that as a *good thing* but if we still insist any hobbyist should be able to run a server with solid events indexer, best apis, etc this will block much progress down this road. Unless we just provide eg docker compose for hobbyist that provides a recommended setup including external services (external dbs, messaging queues, api servers, etc)

Basically requiring 3+ services to run and work together, which I think is needed for any decently functional web tech, will quickly hit alarms of "hobbyist can't run it". I just want to highlight this tension early on and seek a resolution of it before more time gets spent on ADRs or worse implementation, which will be blocked on such grounds

Reply:
I think that this is a pretty fundamental tension within Tendermint that I'd be interested in seeing resolved. Resolving that seems like a rather large task and I'm not certain how to begin it. I would say that I am firmly on the side that hobbyists should not be considered the primary audience for running Tendermint. This tension comes up fairly frequently. 

For example, it comes up when discussing logging within Tendermint. Operators sometimes complain that the logging has a tendency to 'quickly fill their terminal' so tools to filter logging should be built into Tendermint. While I'm sympathetic to not simply spamming the logs with useless nonsense, it does appear to me that common devops practices like running an external log indexer are not necessarily common for operators. Similar tension arises for me when I discuss how to access prometheus metrics with operators who don't always seem to be running a prom instance. If we want to make this shift in direction which, as @ethanfrey deftly points out, will allow faster development, we may also have to do a decent amount of documentation / explanation / evangelism to clarify why this is preferred.